tensorflow/core/grappler/optimizers/auto_mixed_precision.cc

```diff
@@ -1392,35 +1340,144 @@ Status AutoMixedPrecisionImpl::Optimize() {
   return Status::OK();
 }
 
-// Finds data structure object ops (e.g., StackV2) and the sets of nodes that
-// write (e.g., StackPushV2) and read (e.g., StackPopV2) from them.
-Status AutoMixedPrecisionImpl::AddDataStructureOpsToMap(
-    const absl::flat_hash_set<string>& data_structure_ops,
-    TypeAttrId data_structure_type_attr,
-    const absl::flat_hash_map<string, TypeAttrId>& write_ops,
-    const absl::flat_hash_map<string, TypeAttrId>& read_ops,
-    DataStructureOpsMap* object_clients_map) const {
-  for (const NodeDef& node : graph_->node()) {
-    const auto write_iter = write_ops.find(node.op());
-    const auto read_iter = read_ops.find(node.op());
-    bool is_writer = write_iter != write_ops.end();
-    bool is_reader = read_iter != read_ops.end();
-    if (is_writer || is_reader) {
-      const NodeDef* object_node = GetTailOfChain(node, data_structure_ops);
-      if (!object_node) {
-        return errors::FailedPrecondition(
-            "No data structure op found upstream of ", node.op(), " node ",
-            node.name());
+// If node is a Tensor List op with a float32 data type attribute then this
+// returns a pointer to the NodeTypeId representing that type attribute. In
+// all other cases this returns nullptr.
+const NodeTypeId* AutoMixedPrecisionImpl::GetTensorListFloat32NodeTypeId(
+    const NodeDef& node) const {
+  if (!IsTensorListOp(node.op())) return nullptr;
+  for (const TypeAttrId& type_attr : node_type_map_.GetTypeAttrs(node)) {
+    const NodeTypeId* node_type =
+        graph_type_view_.GetNode(node.name(), type_attr);
+    if (node_type && node_type->type_attr.fixed_type == DT_INVALID &&
+        node_type->type_attr.type_index == TypeAttrId::kSingleType &&
+        IsFloat32(*node_type)) {
+      return node_type;
+    }
+  }
+  return nullptr;
+}
+
+bool AutoMixedPrecisionImpl::IsSourceOrSinkOp(const string& op) const {
```

Comment:
Talked to @rmlarsen offline. We don't think there's a great way of checking this. It might be possible to have some heuristics, e.g. an op that takes a DT_VARIANT as input and has no outputs, or vice versa. But I'm not sure if it's worth implementing this.

Having Python tests to catch this, in addition to the C++ tests, would be nice in case we change which op is used on the Python side. But this doesn't need to be done in this PR.

Reply:
Added a TODO to the Python tests.